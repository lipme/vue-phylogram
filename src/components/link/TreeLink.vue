// Copyright 2020 [LIPM]
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

// http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
//    limitations under the License.
<template>
<g class="link" >
  <title v-if="support">Support value : {{support}}</title>
  <path
    title="link"
    :id="id"
    class="link"
    :d="path"
    fill="none"
    :stroke="stroke"
    :stroke-width="strokeWidth"
    :stroke-dasharray="dashAttribute"
  >
    <animate
      attributeName="stroke-width"
      attributeType="XML"
      from="0"
      :to="strokeWidth"
      begin="0s"
      dur="2s"
    />
  </path>
  <!-- <text
      class="support"
      :transform="transformSupport"
      :font-size="supportFontSize"
      :fill="supportColor"
      lengthAdjust="spacingAndGlyphs"
    >
    {{this.support}}
  </text> -->
  <text :font-size="supportFontSize" dy="2px" dx="2px">
     <textPath :xlink:href="xRefId" startOffset="30%">{{this.support}}</textPath>
  </text>
    </g>

</template>

<script>
import * as d3Shape from 'd3-shape'

export default {
  props: {
    id: {
      type: String, default: 'link'
    },
    stroke: {
      type: String,
      default: 'black'
    },
    strokeWidth: {
      type: Number,
      default: 0.3
    },
    rightAngle: {
      type: Boolean,
      default: true
    },
    circular: {
      type: Boolean,
      default: false
    },
    source: {
      type: Object,
      default: () => {
        return {
          x: 20,
          y: 10
        }
      }
    },
    target: {
      type: Object,
      default: () => {
        return {
          x: 40,
          y: 60
        }
      }
    },
    dashed: {
      type: Boolean,
      default: false
    },
    support: {
      type: String,
      default: ''
    },
    supportColor: {
      type: String,
      default: 'black'
    },
    supportFontSize: {
      type: Number,
      default: 12.0
    }
  },
  computed: {
    xRefId () {
      return '#' + this.id
    },
    path () {
      if (this.rightAngle === false) {
        if (this.circular === false) {
          return this.linkHorizontal()
        } else {
          return this.linkRadial()
        }
      }

      const pair = { source: this.source, target: this.target }
      if (this.circular === false) {
        return this.rightAngleDiagonal()(pair)
      } else {
        return this.radialRightAngleDiagonal()(pair)
      }
    },
    dashAttribute () {
      const step = this.strokeWidth * 5
      return this.dashed ? step + ',' + step : ''
    },
    transformSupport () {
      return 'translate(' + ((this.source.y + this.target.y) / 2) + ',' + ((this.target.x - this.supportFontSize / 2)) + ')'
    }
  },
  methods: {
    rightAngleDiagonal () {
      /**
       * From d3.phylogram.js
       * https://gist.github.com/biovisualize/5437184
       */
      let projection = function (d) {
        return [d.y, d.x]
      }

      let path = function (pathData) {
        return 'M' + pathData[0] + ' ' + pathData[1] + ' ' + pathData[2]
      }

      function diagonal (diagonalPath, i) {
        const source = diagonalPath.source
        const target = diagonalPath.target
        let pathData = [source, { x: target.x, y: source.y }, target]
        pathData = pathData.map(projection)
        return path(pathData)
      }

      diagonal.projection = function (x) {
        if (!arguments.length) return projection
        projection = x
        return diagonal
      }

      diagonal.path = function (x) {
        if (!arguments.length) return path
        path = x
        return diagonal
      }

      return diagonal
    },
    radialRightAngleDiagonal () {
      const elt = this

      return this.rightAngleDiagonal()
        .path(function (pathData) {
          const src = pathData[0]
          const mid = pathData[1]
          const dst = pathData[2]
          const radius = Math.sqrt(src[0] * src[0] + src[1] * src[1])
          const srcAngle = elt.coordinateToAngle(src, radius)
          const midAngle = elt.coordinateToAngle(mid, radius)
          const clockwise =
            Math.abs(midAngle - srcAngle) > Math.PI
              ? midAngle <= srcAngle
              : midAngle > srcAngle
          const rotation = 0
          const largeArc = 0
          const sweep = clockwise ? 0 : 1
          return (
            'M' +
            src +
            ' ' +
            'A' +
            [radius, radius] +
            ' ' +
            rotation +
            ' ' +
            largeArc +
            ',' +
            sweep +
            ' ' +
            mid +
            'L' +
            dst
          )
        })
        .projection(function (d) {
          const r = d.y
          const a = ((d.x - 90) / 180) * Math.PI
          return [r * Math.cos(a), r * Math.sin(a)]
        })
    },
    coordinateToAngle (coord, radius) {
      const wholeAngle = 2 * Math.PI
      const quarterAngle = wholeAngle / 4

      const coordQuad =
        coord[0] >= 0 ? (coord[1] >= 0 ? 1 : 2) : coord[1] >= 0 ? 4 : 3
      const coordBaseAngle = Math.abs(Math.asin(coord[1] / radius))

      // Since this is just based on the angle of the right triangle formed
      // by the coordinate and the origin, each quad will have different
      // offsets
      let coordAngle

      switch (coordQuad) {
        case 1:
          coordAngle = quarterAngle - coordBaseAngle
          break
        case 2:
          coordAngle = quarterAngle + coordBaseAngle
          break
        case 3:
          coordAngle = 2 * quarterAngle + quarterAngle - coordBaseAngle
          break
        case 4:
          coordAngle = 3 * quarterAngle + coordBaseAngle
      }
      return coordAngle
    },
    linkHorizontal () {
      const o = {
        source: this.source,
        target: this.target
      }

      const link = d3Shape
        .linkHorizontal()
        .x(function (d) {
          return d.y
        })
        .y(function (d) {
          return d.x
        })

      return link(o)
    },
    linkRadial () {
      const o = {
        source: this.source,
        target: this.target
      }

      const link = d3Shape
        .linkRadial()
        .angle(function (d) {
          return (d.x / 180) * Math.PI
        })
        .radius(function (d) {
          return d.y
        })

      return link(o)
    }
  }
}
</script>

<style scoped>
path:hover {
  transition: all 0.5s;
  stroke: red;
}

</style>
