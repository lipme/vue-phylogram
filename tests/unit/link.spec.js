// Copyright 2020 [LIPM]
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at

// http://www.apache.org/licenses/LICENSE-2.0

// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
//    limitations under the License.
import { shallowMount } from '@vue/test-utils'
import TreeLink from '@/components/link'

describe('TreeLink component', () => {
  it('default link renders correctly', () => {
    const comp = shallowMount(TreeLink, {
      propsData: {
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('color set correctly', () => {
    const comp = shallowMount(TreeLink, {
      propsData: {
        stroke: 'red'
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('width set correctly', () => {
    const comp = shallowMount(TreeLink, {
      propsData: {
        strokeWidth: 1
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('smooth mode set correctly', () => {
    const comp = shallowMount(TreeLink, {
      propsData: {
        rightAngle: false
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('circular mode set correctly', () => {
    const comp = shallowMount(TreeLink, {
      propsData: {
        circular: true
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('circular mode and smooth mode set correctly', () => {
    const comp = shallowMount(TreeLink, {
      propsData: {
        circular: true,
        rightAngle: false
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('source set correctly', () => {
    const comp = shallowMount(TreeLink, {
      propsData: {
        source: {
          x: 2,
          y: 1
        }
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('target set correctly', () => {
    const comp = shallowMount(TreeLink, {
      propsData: {
        target: {
          x: 4,
          y: 6
        }
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
  it('dashed mode set correctly', () => {
    const comp = shallowMount(TreeLink, {
      propsData: {
        dashed: true
      }
    })

    expect(comp.element).toMatchSnapshot()
  })
})
