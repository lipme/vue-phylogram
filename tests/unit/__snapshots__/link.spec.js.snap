// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeLink component circular mode and smooth mode set correctly 1`] = `
<path
  class="link"
  d="M3.4202014332566884,-9.396926207859083C11.970705016398409,-32.88924172750679,22.49756633902888,-26.81155550916423,38.56725658119236,-45.96266658713868"
  fill="none"
  stroke="black"
  stroke-dasharray=""
  stroke-width="0.3"
>
  <animate
    attributeName="stroke-width"
    attributeType="XML"
    begin="0s"
    dur="2s"
    from="0"
    to="0.3"
  />
</path>
`;

exports[`TreeLink component circular mode set correctly 1`] = `
<path
  class="link"
  d="M3.4202014332566884,-9.396926207859083 A10,10 0 0,1 6.427876096865393,-7.66044443118978L38.56725658119236,-45.96266658713868"
  fill="none"
  stroke="black"
  stroke-dasharray=""
  stroke-width="0.3"
>
  <animate
    attributeName="stroke-width"
    attributeType="XML"
    begin="0s"
    dur="2s"
    from="0"
    to="0.3"
  />
</path>
`;

exports[`TreeLink component color set correctly 1`] = `
<path
  class="link"
  d="M10,20 10,40 60,40"
  fill="none"
  stroke="red"
  stroke-dasharray=""
  stroke-width="0.3"
>
  <animate
    attributeName="stroke-width"
    attributeType="XML"
    begin="0s"
    dur="2s"
    from="0"
    to="0.3"
  />
</path>
`;

exports[`TreeLink component dashed mode set correctly 1`] = `
<path
  class="link"
  d="M10,20 10,40 60,40"
  fill="none"
  stroke="black"
  stroke-dasharray="1.5,1.5"
  stroke-width="0.3"
>
  <animate
    attributeName="stroke-width"
    attributeType="XML"
    begin="0s"
    dur="2s"
    from="0"
    to="0.3"
  />
</path>
`;

exports[`TreeLink component default link renders correctly 1`] = `
<path
  class="link"
  d="M10,20 10,40 60,40"
  fill="none"
  stroke="black"
  stroke-dasharray=""
  stroke-width="0.3"
>
  <animate
    attributeName="stroke-width"
    attributeType="XML"
    begin="0s"
    dur="2s"
    from="0"
    to="0.3"
  />
</path>
`;

exports[`TreeLink component smooth mode set correctly 1`] = `
<path
  class="link"
  d="M10,20C35,20,35,40,60,40"
  fill="none"
  stroke="black"
  stroke-dasharray=""
  stroke-width="0.3"
>
  <animate
    attributeName="stroke-width"
    attributeType="XML"
    begin="0s"
    dur="2s"
    from="0"
    to="0.3"
  />
</path>
`;

exports[`TreeLink component source set correctly 1`] = `
<path
  class="link"
  d="M1,2 1,40 60,40"
  fill="none"
  stroke="black"
  stroke-dasharray=""
  stroke-width="0.3"
>
  <animate
    attributeName="stroke-width"
    attributeType="XML"
    begin="0s"
    dur="2s"
    from="0"
    to="0.3"
  />
</path>
`;

exports[`TreeLink component target set correctly 1`] = `
<path
  class="link"
  d="M10,20 10,4 6,4"
  fill="none"
  stroke="black"
  stroke-dasharray=""
  stroke-width="0.3"
>
  <animate
    attributeName="stroke-width"
    attributeType="XML"
    begin="0s"
    dur="2s"
    from="0"
    to="0.3"
  />
</path>
`;

exports[`TreeLink component width set correctly 1`] = `
<path
  class="link"
  d="M10,20 10,40 60,40"
  fill="none"
  stroke="black"
  stroke-dasharray=""
  stroke-width="1"
>
  <animate
    attributeName="stroke-width"
    attributeType="XML"
    begin="0s"
    dur="2s"
    from="0"
    to="1"
  />
</path>
`;
